[GStreamer] Add helper function to create video conversion elements and enable multi-threading inside

philn opened the pull request:
#### 7ba265366dadf5387c38b44c068575eed2ea8118
<pre>
[GStreamer] Add helper function to create video conversion elements and enable multi-threading inside
<a href="https://bugs.webkit.org/show_bug.cgi?id=302252">https://bugs.webkit.org/show_bug.cgi?id=302252</a>

Reviewed by Xabier Rodriguez-Calvar.

Use videoconvertscale when running on at least GStreamer 1.22 and if the
WEBKIT_GST_USE_VIDEOCONVERT_SCALE environment variable is set to 1. Otherwise build a bin containing
a videoscale and videoconvert element. Multi-threading is also now enabled in videoconvert, which
should improve performance, specially when processing video frames with large dimensions.

* Source/WebCore/platform/graphics/gstreamer/GStreamerCommon.cpp:
(WebCore::createVideoConvertScaleElement):
* Source/WebCore/platform/graphics/gstreamer/GStreamerCommon.h:
* Source/WebCore/platform/graphics/gstreamer/GStreamerVideoFrameConverter.cpp:
(WebCore::GStreamerVideoFrameConverter::Pipeline::Pipeline):
* Source/WebCore/platform/gstreamer/VideoEncoderPrivateGStreamer.cpp:
(videoEncoderSetEncoder):
* Source/WebCore/platform/mediastream/gstreamer/GStreamerVideoCapturer.cpp:
(WebCore::GStreamerVideoCapturer::createConverter):

Canonical link: <a href="https://commits.webkit.org/302902@main">https://commits.webkit.org/302902@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/2ee3ecddce3c641ffea0468b5fd1a0378b9bfb67

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/130620 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/159/builds/2891 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/41575 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/138040 "Built successfully") | [  ~~🛠 win~~](https://ews-build.webkit.org/#/builders/59/builds/82241 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/132491 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/155/builds/2903 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/156/builds/2783 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/99512 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/59/builds/82241 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/133567 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/162/builds/2105 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/116959 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/80219 "Passed tests") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/154/builds/2026 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/35093 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/81297 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/110598 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/35595 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/140519 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/153/builds/2681 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/161/builds/2431 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/108017 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/160/builds/2725 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/113303 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/107949 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/27463 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/164/builds/2066 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/31730 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/55660 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/158/builds/2751 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/66140 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/157/builds/2570 "Built successfully") | | | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/163/builds/2772 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/152/builds/2677 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

calvaris commented on Source/WebCore/platform/graphics/gstreamer/GStreamerCommon.cpp:
```diff
@@ -2108,6 +2108,43 @@ GstStateChangeReturn gstElementLockAndSetState(GstElement* element, GstState sta
+            return nullptr;
+
+        // Enable multi-threading in the converter.
+        g_object_set(videoConvert.get(), "n-threads", static_cast<unsigned>(0), nullptr);
```
You can probably use `0U` instead of casting like this.

calvaris commented on Source/WebCore/platform/graphics/gstreamer/GStreamerCommon.cpp:
```diff
@@ -2108,6 +2108,43 @@ GstStateChangeReturn gstElementLockAndSetState(GstElement* element, GstState sta
+        return nullptr;
+
+    // Enable multi-threading in the converter.
+    g_object_set(videoConvert, "n-threads", static_cast<unsigned>(0), nullptr);
```
ditto

calvaris commented on Source/WebCore/platform/graphics/gstreamer/GStreamerVideoFrameConverter.cpp:
```diff
@@ -53,11 +53,11 @@ GStreamerVideoFrameConverter::Pipeline::Pipeline(Type type)
     case Type::SystemMemory: {
-        auto videoconvert = makeGStreamerElement("videoconvert"_s);
-        auto videoscale = makeGStreamerElement("videoscale"_s);
+        auto videoConvert = createVideoConvertScaleElement();
```
This should be named as videoConvertScale, right?

calvaris commented on Source/WebCore/platform/gstreamer/VideoEncoderPrivateGStreamer.cpp:
```diff
@@ -313,27 +313,11 @@ static bool videoEncoderSetEncoder(WebKitVideoEncoder* self, EncoderId encoderId
-        videoConvert = makeGStreamerElement("videoconvert"_s);
-        if (!videoConvert)
-            return false;
+    auto videoConvert = createVideoConvertScaleElement();
```
ditto

calvaris commented on Source/WebCore/platform/mediastream/gstreamer/GStreamerVideoCapturer.cpp:
```diff
@@ -128,15 +128,17 @@ GstElement* GStreamerVideoCapturer::createConverter()
         return nullptr;
     }
 
+    auto videoConvert = createVideoConvertScaleElement();
```
ditto

philn committed:
[GStreamer] Add helper function to create video conversion elements and enable multi-threading inside
https://bugs.webkit.org/show_bug.cgi?id=302252

Reviewed by Xabier Rodriguez-Calvar.

Use videoconvertscale when running on at least GStreamer 1.22 and if the
WEBKIT_GST_USE_VIDEOCONVERT_SCALE environment variable is set to 1. Otherwise build a bin containing
a videoscale and videoconvert element. Multi-threading is also now enabled in videoconvert, which
should improve performance, specially when processing video frames with large dimensions.

* Source/WebCore/platform/graphics/gstreamer/GStreamerCommon.cpp:
(WebCore::createVideoConvertScaleElement):
* Source/WebCore/platform/graphics/gstreamer/GStreamerCommon.h:
* Source/WebCore/platform/graphics/gstreamer/GStreamerVideoFrameConverter.cpp:
(WebCore::GStreamerVideoFrameConverter::Pipeline::Pipeline):
* Source/WebCore/platform/gstreamer/VideoEncoderPrivateGStreamer.cpp:
(videoEncoderSetEncoder):
* Source/WebCore/platform/mediastream/gstreamer/GStreamerVideoCapturer.cpp:
(WebCore::GStreamerVideoCapturer::createConverter):

Canonical link: https://commits.webkit.org/302902@main
